### PR TITLE
Distributed define_spmd_block

### PR DESCRIPTION
**Pull request by atrantan**
This patch tries to implement a distributed version of `define_spmd_block()`.Here are the following changes:
- local `define_spmd_block()` and local `spmd_block()` have been moved to `hpx::lcos::local`
- `hpx::parallel::v2::spmd_block` is now defined as a type alias
- `hpx::parallel::v2::define_spmd_block()` is now wrapping a call to `hpx::lcos::local::define_spmd_block()`
- global `define_spmd_block()` and global `spmd_block()`  (based on `broadcast` and `hpx::lcos::barrier` ) added and located in `hpx::lcos`

### CODE REVIEW

File: hpx/lcos/spmd_block.hpp [hunk 1]
```diff
@@ -0,0 +1,275 @@
...
+
+                static_assert(std::is_same<spmd_block, first_type>::value,
+                    "define_spmd_block() needs a lambda that " \
+                    "has at least a spmd_block as 1st argument");
+
+                spmd_block block(name_, num_images_, image_id);
+
+                int * dummy = nullptr;
+                hpx::util::invoke(
+                    reinterpret_cast<const F&>(*dummy),
```
Comment (sithhell) [inline hpx/lcos/spmd_block.hpp]: How should this work? invoke will certainly dereference a nullptr. This is UB.

File: hpx/lcos/spmd_block.hpp [hunk 2]
```diff
@@ -0,0 +1,275 @@
...
+        using helper_type =
+            hpx::lcos::detail::spmd_block_helper_action<
+               ftype, typename std::decay<Args>::type...>;
+
+        using helper_action_type =
+            typename hpx::actions::make_action<
+                decltype( &helper_type::call ), &helper_type::call >::type;
+
+        // Initialize f at compile-time in case when f is a lambda function
+        hpx::lcos::detail::initialize_if_lambda(f);
```
Comment (sithhell) [inline hpx/lcos/spmd_block.hpp]: This doesn't really work. It only might work if `f` has no state. This limits the usefulness of `spmd_block` tremenedously.The nullptr dereference and that you don't transfer the callable to the remote is just calling for trouble.

File: hpx/lcos/spmd_block.hpp (hunk 1 above)
Comment (atrantan) [inline hpx/lcos/spmd_block.hpp]: Lambda overload removed in 4657f0e1ae4edfd6f4fb640cd6a67f25426b80f4

File: hpx/lcos/spmd_block.hpp (hunk 2 above)
Comment (atrantan) [inline hpx/lcos/spmd_block.hpp]: Fixed in 4657f0e1ae4edfd6f4fb640cd6a67f25426b80f4

File: hpx/parallel/spmd_block.hpp [hunk 3]
```diff
@@ -7,190 +7,59 @@
...
-    }
+    using spmd_block = hpx::lcos::local::spmd_block;
 
     // Asynchronous version
     template <typename ExPolicy, typename F, typename ... Args,
         typename = typename std::enable_if<
             hpx::parallel::v1::is_async_execution_policy<ExPolicy>::value>::type
         >
-    std::vector<hpx::future<void>>
+    inline std::vector<hpx::future<void>>
```
Comment (Naios) [inline hpx/parallel/spmd_block.hpp]: templated functions have `inline` linkage by default

Comment (atrantan) [inline hpx/parallel/spmd_block.hpp]: Fix done in 9d28da96839988f983c3bc5fa975a4a814b1e22c

**Comment by atrantan** [comment]
@sithhell Is the PR now good for U?

File: hpx/lcos/spmd_block.hpp [hunk 4]
```diff
@@ -0,0 +1,211 @@
...
+        {
+            using type = std::false_type;
+        };
+
+        template <typename Arg0, typename... Args>
+        struct extract_first_parameter< hpx::util::tuple<Arg0, Args...> >
+        {
+            using type = typename std::decay<Arg0>::type;
+        };
+    }
```
Comment (sithhell) [inline hpx/lcos/spmd_block.hpp]: This should be factored out and merged with the one defined in `hpx/lcos/local/spmd_block.hpp`

Comment (atrantan) [inline hpx/lcos/spmd_block.hpp]: @sithhell Done in 77ecb48d61b9e956610383a5da5d5c1add2ea9ce, but not sure if I'm covering all the cases and if the proposed approach is good.

**Pull request merged**